typing/jkind.ml

```diff
@@ -383,89 +383,117 @@ let raise ~loc err = raise (Error.User_error (loc, err))
+    let locality = Locality.join (locality t1) (locality t2) in
+    let linearity = Linearity.join (linearity t1) (linearity t2) in
+    let uniqueness = Uniqueness.join (uniqueness t1) (uniqueness t2) in
+    let portability = Portability.join (portability t1) (portability t2) in
+    let contention = Contention.join (contention t1) (contention t2) in
+    let yielding = Yielding.join (yielding t1) (yielding t2) in
+    let externality = Externality.join (externality t1) (externality t2) in
+    let nullability = Nullability.join (nullability t1) (nullability t2) in
+    create ~locality ~linearity ~uniqueness ~portability ~contention ~yielding
+      ~externality ~nullability
+
+  let meet t1 t2 =
+    let locality = Locality.meet (locality t1) (locality t2) in
+    let linearity = Linearity.meet (linearity t1) (linearity t2) in
+    let uniqueness = Uniqueness.meet (uniqueness t1) (uniqueness t2) in
+    let portability = Portability.meet (portability t1) (portability t2) in
+    let contention = Contention.meet (contention t1) (contention t2) in
+    let yielding = Yielding.meet (yielding t1) (yielding t2) in
+    let externality = Externality.meet (externality t1) (externality t2) in
+    let nullability = Nullability.meet (nullability t1) (nullability t2) in
+    create ~locality ~linearity ~uniqueness ~portability ~contention ~yielding
+      ~externality ~nullability
+
+  let less_or_equal t1 t2 =
+    let axis_less_or_equal ~le ~axis a b : Sub_result.t =
+      match le a b, le b a with
+      | true, true -> Equal
+      | true, false -> Less
+      | false, _ -> Not_le [Axis_disagreement axis]
+    in
+    Sub_result.combine
+      (axis_less_or_equal ~le:Locality.le
+         ~axis:(Pack (Modal (Comonadic Areality))) (locality t1) (locality t2))
+    @@ Sub_result.combine
+         (axis_less_or_equal ~le:Uniqueness.le
+            ~axis:(Pack (Modal (Monadic Uniqueness))) (uniqueness t1)
+            (uniqueness t2))
+    @@ Sub_result.combine
+         (axis_less_or_equal ~le:Linearity.le
+            ~axis:(Pack (Modal (Comonadic Linearity))) (linearity t1)
+            (linearity t2))
+    @@ Sub_result.combine
+         (axis_less_or_equal ~le:Contention.le
+            ~axis:(Pack (Modal (Monadic Contention))) (contention t1)
+            (contention t2))
+    @@ Sub_result.combine
+         (axis_less_or_equal ~le:Portability.le
+            ~axis:(Pack (Modal (Comonadic Portability))) (portability t1)
+            (portability t2))
+    @@ Sub_result.combine
+         (axis_less_or_equal ~le:Yielding.le
+            ~axis:(Pack (Modal (Comonadic Yielding))) (yielding t1)
+            (yielding t2))
+    @@ Sub_result.combine
+         (axis_less_or_equal ~le:Externality.le
+            ~axis:(Pack (Nonmodal Externality)) (externality t1)
+            (externality t2))
+    @@ axis_less_or_equal ~le:Nullability.le ~axis:(Pack (Nonmodal Nullability))
+         (nullability t1) (nullability t2)
+
+  let equal t1 t2 =
```

Comment:
Another good place for a pattern-match

Reply:
see also https://github.com/ocaml-flambda/flambda-backend/pull/3605#discussion_r1970422090